core/parser/src/parser/expression/mod.rs

```diff
@@ -73,30 +75,35 @@ macro_rules! expression {
         where
             R: ReadChar
         {
-            type Output = ast::Expression;
+            type Output = FormalParameterListOrExpression;
 
-            fn parse(self, cursor: &mut Cursor<R>, interner: &mut Interner)-> ParseResult<ast::Expression> {
+            fn parse(self, cursor: &mut Cursor<R>, interner: &mut Interner)-> ParseResult<Self::Output> {
 
                 if $goal.is_some() {
                     cursor.set_goal($goal.unwrap());
                 }
 
-                let mut lhs = $lower::new($( self.$low_param ),*).parse(cursor, interner)?;
+                let lhs = $lower::new($( self.$low_param ),*).parse(cursor, interner)?;
+                let mut lhs = match lhs {
+                    FormalParameterListOrExpression::Expression(expression) => expression,
+                    other => return Ok(other)
+                };
```

Comment:
Remark: It might be a bit cleaner to use let-else here.
```suggestion
                let FormalParameterListOrExpression::Expression(mut lhs) = lhs else {
                    return Ok(lhs);
                };
```